ouroboros-consensus-test/bench-storage/Bench/Ouroboros/Consensus/Storage/LedgerDB/HD.hs

```diff
@@ -0,0 +1,412 @@
+                           , NFData (Diff ds k v)
+                           , NFData (Values ds k v)
+                           , NFData (Keys ds k v)
+                           , NFData (SlotNo ds)
+                           ) => NFData (Cmd ds k v)
+
+deriving stock instance ( Show (ds k v)
+                        , Show (Diff ds k v)
+                        , Show (Values ds k v)
+                        , Show (Keys ds k v)
+                        , Show (SlotNo ds)
+                        ) => Show (Cmd ds k v)
+
+deriving stock instance ( Eq (ds k v)
+                        , Eq (Diff ds k v)
+                        , Eq (Values ds k v)
+                        , Eq (Keys ds k v)
+                        , Eq (SlotNo ds)
+                        ) => Eq (Cmd ds k v)
+
+{-------------------------------------------------------------------------------
+  Interpreter
+-------------------------------------------------------------------------------}
+
+-- | Interpretation of @Cmd@s as a fold over an initial diff sequence.
+interpret ::
+     forall ds k v.
+     ( IsDiffSeq ds k v
+     , NFData (Diff ds k v)
+     , NFData (Values ds k v)
+     )
+  => ds k v
+  -> [Cmd ds k v]
+  -> ds k v
+interpret = foldl' go
+  where
+    go :: ds k v -> Cmd ds k v -> ds k v
+    go ds = \case
+      Push sl d     -> push ds sl d
+      -- We @'deepseq'@ the total diff of the left part of the flush result,
+      -- because that is the part that would be written to disk, for which it
+      -- must be fully evaluated. In particular, the performance of computing
+      -- the total diff can be different for each diff sequence implementation.
+      Flush n       -> let (l, r) = flush n ds
+                       in  totalDiff l `deepseq` r
+      -- We do not @'deepseq'@ the right part of the rollback result, because we
+      -- are throwing it away in the case of a rollback.
+      Rollback n bs -> let (l, _r) = rollback n ds
+                       in  pushMany l bs
+      -- We @'deepseq'@ the result of forwarding values through a total diff,
+      -- since forwarding can be implementation-specific in (at least) two ways:
+      -- 1. How the total diff is computed.
+      -- 2. How the change to a specific key-value pair is computed.
+      Forward vs ks -> let vs' = forward vs ks ds
+                       in  vs' `deepseq` ds
+
+{-------------------------------------------------------------------------------
+  Type class for diff sequences
+-------------------------------------------------------------------------------}
+
+-- | The kind of a diff sequence type that is parameterised by the type of keys
+-- and values respectively.
+--
+-- Note: Diffs in this context are diffs for key-value stores (a.k.a. @'Map'@),
+-- not diffs for arbitrary datatypes.
+type DiffSeqKind = Type -> Type -> Type
+
+-- | Common interface to diff sequence implementations.
+class IsDiffSeq (ds :: DiffSeqKind) k v where
```

Comment:
If I understand correctly, the deps are "necessary" because `ds` is annotated with a kind. I do prefer the current defintion of the type class over one of these alternatives:
```haskell
class IsDiffSeq (ds :: Type) k v where
  type Diff ds k v = ...
  ...
  
class IsDiffSeq (ds :: Type) where
  type Diff ds k v = ...
  ...
```